### ISSUE TITLE
Using autowrap to support explicit data type conversion in wrapper functions.

### ISSUE THREAD

**Issue by 24sharkS** [OpenMS/OpenMS#4776]
In regards to the [card](https://github.com/OpenMS/OpenMS/projects/31#card-39205182).
- How can autowrap help to evaluate the wrapper functions where explicit conversion of arguments is necessary. (e.g. load function of IdXMLFile class, as the peptide and protein lists are modified)
- From the result of parsing a pxd file, after analyzing the information related to the data type of arguments, we may create a helper function to facilitate the type conversions inside our wrapper methods.

**Comment by timosachsenberg** [comment]
@hroest 
is there a way in autowrap to determine that a parameter that is passed by reference e.g.,
https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/pxds/IdXMLFile.pxd#L13
is non-const in the underlying C++ implementation?
I.e. like here https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/FORMAT/IdXMLFile.h#L85

**Comment by timosachsenberg** [comment]
For example https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/pxds/Acquisition.pxd#L16
contains the const specifier in the pxd file so maybe this is optional but possible.

**Comment by timosachsenberg** [comment]
Hi @uweschmitt ,
If you find the time it would be great to comment if the way (by reference, by const reference, by value) parameters are passed is available at some point in autowrap. P.S.: We are evaluating if autowrap could also be used to generate R wrapper (e.g., with reticulate).

**Comment by hroest** [comment]
There is a bit of non-compliance going on here: in principle autowrap can distinguish between passing by reference and passing by value and will generate different code for each:

- passing by value will mean that input gets passed into the function (no more work)
- passing by value means that everything passed into the function needs to be read again after the function has run and copied back into Python data structures

imagine you have a 
```
void foobar(std::vector<int>& myList)
```

function. What autowrap needs to do is take a Python list, convert it to `std::vector` and pass it into the function, then look at the result and copy the result back into a Python list.

Now the main problem was that for a long time autowtrap did *not* distinguish between ref and const-ref functions, therefore all functions that were optimized for speed with const-ref still triggered autowrap to create the code for ref (copy into C++ struct, copy back). This is inefficient and we have thus opted to annotated const-ref functions sometimes as pass by value functions. Since https://github.com/uweschmitt/autowrap/pull/70 however, autowrap does understand const-ref and this trick is no longer necessary, therefore autowrap can properly distinguish between: i) pass by value ii) pass by ref and iii) pass by const-ref.

In conclusion: due to historic reasons, some const-ref parameters are annotated as pass-by-value in autowrap but nowadays it understands all three models.
